pkg/assembler/backends/inmem/pkg.go

```diff
@@ -421,20 +436,58 @@ func (c *demoClient) IngestPackage(ctx context.Context, input model.PkgInputSpec
 	return c.buildPackageResponse(collectedVersion.id, nil)
 }
 
-func duplicatePkgVer(versions pkgVersionList, input model.PkgInputSpec) (bool, *pkgVersionNode) {
-	for _, v := range versions {
-		if noMatchInput(input.Version, v.version) {
-			continue
-		}
-		if noMatchInput(input.Subpath, v.subpath) {
-			continue
-		}
-		if !reflect.DeepEqual(v.qualifiers, getQualifiersFromInput(input.Qualifiers)) {
-			continue
+// hash the canonical representation of a version
+func hashPkgVersionNode(version *pkgVersionNode) (string, error) {
+	if version == nil {
+		return "", fmt.Errorf("version is nil")
+	}
+	return hashVersionHelper(version.version, version.subpath, version.qualifiers)
+}
+
+// hash the canonical representation of a version
+func hashPkgInputSpecVersion(input model.PkgInputSpec) (string, error) {
+	qualifiers := getQualifiersFromInput(input.Qualifiers)
+	return hashVersionHelper(nilToEmpty(input.Version), nilToEmpty(input.Subpath), qualifiers)
+}
+
+func hashVersionHelper(version string, subpath string, qualifiers map[string]string) (string, error) {
+	hash := sha256.New()
+
+	_, err := hash.Write([]byte(version))
+	if err != nil {
+		return "", fmt.Errorf("Error writing when computing the hash of a version: %v", err)
+	}
+	_, err = hash.Write([]byte(subpath))
+	if err != nil {
+		return "", fmt.Errorf("Error writing when computing the hash of a version: %v", err)
+	}
+
+	qualifierSlice := []string{}
+	for key, value := range qualifiers {
+		qualifierSlice = append(qualifierSlice, fmt.Sprintf("%s:%s", key, value))
+	}
+	slices.Sort(qualifierSlice)
+	for _, qualifier := range qualifierSlice {
+		_, err = hash.Write([]byte(qualifier))
+		if err != nil {
+			return "", fmt.Errorf("Error writing when computing the hash of a version: %v", err)
 		}
-		return true, v
 	}
-	return false, nil
+
+	digest := hash.Sum([]byte{})
+	return fmt.Sprintf("%q", digest), nil
+}
+
+func duplicatePkgVer(versions pkgVersionMap, input model.PkgInputSpec) (bool, *pkgVersionNode, error) {
+	digest, err := hashPkgInputSpecVersion(input)
+	if err != nil {
+		return false, nil, err
+	}
+
+	if version, ok := versions[digest]; ok {
+		return true, version, nil
+	}
+	return false, nil, nil
 }
 
 // Query Package
```

Comment:
I think we can also leverage the map in the query in the query flow (i.e. in `Packages` flow in `buildPkgName`), we may want to check the flow in which `IsDependency` and `IsOccurrence` and other evidence trees lookup packages as well..

If you'd like we can do this as a separate PR as well. Just note it in the original issue.

Reply:
Sounds good, I will follow up with another PR